File: benchmark/languages/c/work/day01/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main(int argc, char *argv[]) {
    int part = 0;
    if (argc >= 2) part = atoi(argv[1]);

    FILE *f = fopen("input.txt", "r");
    if (!f) { fprintf(stderr, "Cannot open input.txt\n"); return 1; }

    char line[256];
    long long part1 = 0, part2 = 0;
    int pos = 50; // dial starts at 50

    while (fgets(line, sizeof(line), f)) {
        char dir;
        int dist;
        if (sscanf(line, " %c%d", &dir, &dist) != 2) continue;

        // Part 2: count passes through 0 during this rotation
        // For L: we step from pos, going down (mod 100)
        //   Step k (1..dist) lands on (pos - k + 10000) % 100
        //   Hits 0 when (pos - k) % 100 == 0, i.e. k % 100 == pos % 100
        //   First hit: if pos == 0, first_k = 100; else first_k = pos
        // For R: we step from pos, going up (mod 100)
        //   Step k (1..dist) lands on (pos + k) % 100
        //   Hits 0 when (pos + k) % 100 == 0, i.e. k % 100 == (100 - pos) % 100
        //   First hit: if pos == 0, first_k = 100; else first_k = 100 - pos
        long long clicks_through_zero = 0;
        if (dir == 'L') {
            int first_k = (pos == 0) ? 100 : pos;
            if (dist >= first_k) {
                // count k = first_k, first_k+100, first_k+200, ... <= dist
                clicks_through_zero = (long long)(dist - first_k) / 100 + 1;
            }
        } else { // R
            int first_k = (pos == 0) ? 100 : (100 - pos);
            if (dist >= first_k) {
                clicks_through_zero = (long long)(dist - first_k) / 100 + 1;
            }
        }
        part2 += clicks_through_zero;

        // Compute new position
        if (dir == 'L') {
            pos = ((pos - dist) % 100 + 100) % 100;
        } else {
            pos = (pos + dist) % 100;
        }

        // Part 1: count if final position is 0
        if (pos == 0) {
            part1++;
        }
    }

    fclose(f);

    // Part 2 total = clicks through 0 during rotation (including landing on 0)
    // But wait: part2 as computed counts landing on 0 as well (since we count step k=dist too if it hits 0)
    // part1 counts only end-of-rotation zeros
    // The problem says part2 = total times dial points at 0 during any click
    // That includes the final click landing on 0 (which part1 counts)
    // So part2 already includes everything (both mid-rotation and end-of-rotation zeros)
    // Let's verify with example: expected part1=3, part2=6

    if (part == 0 || part == 1) {
        printf("Part 1: %lld\n", part1);
    }
    if (part == 0 || part == 2) {
        printf("Part 2: %lld\n", part2);
    }

    return 0;
}

File: benchmark/languages/go/work/day01/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	part := 0
	if len(os.Args) > 1 {
		switch os.Args[1] {
		case "1":
			part = 1
		case "2":
			part = 2
		}
	}

	f, err := os.Open("input.txt")
	if err != nil {
		panic(err)
	}
	defer f.Close()

	type Rotation struct {
		dir string
		dist int
	}

	var rotations []Rotation
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if len(line) == 0 {
			continue
		}
		dir := string(line[0])
		dist, err := strconv.Atoi(line[1:])
		if err != nil {
			panic(err)
		}
		rotations = append(rotations, Rotation{dir, dist})
	}

	// Part 1: count times dial ends at 0 after each rotation
	part1 := func() int {
		pos := 50
		count := 0
		for _, r := range rotations {
			if r.dir == "L" {
				pos = ((pos - r.dist) % 100 + 100) % 100
			} else {
				pos = (pos + r.dist) % 100
			}
			if pos == 0 {
				count++
			}
		}
		return count
	}

	// Part 2: count every click that lands on 0 during any rotation
	// Going Left from pos by d clicks: positions are (pos-1)%100, (pos-2)%100, ..., (pos-d)%100
	// We want count of i in [1..d] where (pos - i) % 100 == 0
	// i.e. (pos - i) ≡ 0 mod 100 => i ≡ pos mod 100
	// So first such i = pos (if pos != 0) or 100 (if pos == 0)
	// Then i = pos, pos+100, pos+200, ... up to d
	// If pos == 0, first i = 100
	// Count = floor((d - first_i) / 100) + 1 if first_i <= d, else 0
	//
	// Going Right from pos by d clicks: positions are (pos+1)%100, ..., (pos+d)%100
	// We want count of i in [1..d] where (pos + i) % 100 == 0
	// (pos + i) ≡ 0 mod 100 => i ≡ -pos ≡ (100 - pos) mod 100
	// first_i = (100 - pos) % 100; if first_i == 0, first_i = 100
	// Count = floor((d - first_i) / 100) + 1 if first_i <= d, else 0

	countZerosDuring := func(pos, dist int, dir string) int {
		var firstI int
		if dir == "L" {
			if pos == 0 {
				firstI = 100
			} else {
				firstI = pos
			}
		} else { // R
			firstI = (100 - pos) % 100
			if firstI == 0 {
				firstI = 100
			}
		}
		if firstI > dist {
			return 0
		}
		return (dist-firstI)/100 + 1
	}

	part2 := func() int {
		pos := 50
		count := 0
		for _, r := range rotations {
			count += countZerosDuring(pos, r.dist, r.dir)
			if r.dir == "L" {
				pos = ((pos - r.dist) % 100 + 100) % 100
			} else {
				pos = (pos + r.dist) % 100
			}
		}
		return count
	}

	if part == 0 || part == 1 {
		fmt.Printf("Part 1: %d\n", part1())
	}
	if part == 0 || part == 2 {
		fmt.Printf("Part 2: %d\n", part2())
	}
}
